### PR TITLE
airframe-sql: Fix SQL generation from resolved plan

### PR DESCRIPTION
**Pull request by takezoe**

### CODE REVIEW

File: airframe-sql/src/test/scala/wvlet/airframe/sql/analyzer/TypeResolverTest.scala [hunk 1]
```diff
@@ -75,23 +75,46 @@ class TypeResolverTest extends AirSpec {
...
+    trace(s"original sql:\n${sql}")
+    trace(s"original plan:\n${SQLParser.parse(sql).pp}")
+
+    val resolvedPlan = resolvePlan(sql, rules)
+    val resolvedSql  = generateSql(resolvedPlan)
+    trace(s"new plan:\n${resolvedPlan.pp}")
+    trace(s"new sql:\n${resolvedSql}")
+
+    // Round-trip plan should be able to be resolved
+    resolvePlan(resolvedSql, rules)
```
Comment (takezoe) [inline airframe-sql/src/test/scala/wvlet/airframe/sql/analyzer/TypeResolverTest.scala:87]: Check if sql generated from resolved plan can be resolved.

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/Expression.scala [hunk 2]
```diff
@@ -313,23 +313,40 @@ object Expression {
...
+            case s: SingleColumn      => s.matchesWith(columnName)
+            case _                    => false
+          }
         case _ => false
       }
     }
   }
   case class MultiColumn(
       inputs: Seq[Expression],
+      name: Option[String],
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/Expression.scala:349]: Added `name` to `MultiColumn` for SQL generation.

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala [hunk 3]
```diff
@@ -795,8 +825,10 @@ object LogicalPlan {
     override def inputAttributes: Seq[Attribute] = Seq.empty // TODO
     override def outputAttributes: Seq[Attribute] = {
       columns.map {
-        case a: ArrayConstructor => SingleColumn(MultiColumn(a.values, None), None, None, a.nodeLocation)
-        case other               => SingleColumn(other, None, None, other.nodeLocation)
+        case _: ArrayConstructor =>
+          ResolvedAttribute(UUID.randomUUID().toString, DataType.AnyType, None, Nil, None) // TODO data type
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala:829]: Removed holding all values in outputAttributes of `Unnest`, but `dataType` cannot be resolved.

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/ResolvedPlan.scala [hunk 4]
```diff
@@ -86,22 +86,16 @@ case class ResolvedAttribute(
...
-    }
-  }
-
   override def toString = {
     (qualifier, sourceColumns) match {
       case (Some(q), columns) if columns.nonEmpty =>
         columns
           .map(_.fullName)
-          .mkString(s"${q},${name}:${dataType} <- ", ", ", "")
+          .mkString(s"${q},${name}:${dataType} <- [", ", ", "]")
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/ResolvedPlan.scala:94]: Just for readability of printed plan

File: airframe-sql/src/main/scala/wvlet/airframe/sql/parser/SQLGenerator.scala [hunk 5]
```diff
@@ -219,7 +219,11 @@ object SQLGenerator extends LogSupport {
         }
       case Join(joinType, left, right, cond, _) =>
         val l = printRelation(left)
-        val r = printRelation(right)
+        val r = right match {
+          case _: Selection    => s"(${printRelation(right)})"
+          case _: SetOperation => s"(${printRelation(right)})"
+          case _               => printRelation(right)
+        }
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/parser/SQLGenerator.scala:226]: Parenthesis is necessary for right-hand of JOIN.

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala [hunk 6]
```diff
@@ -698,7 +709,11 @@ object LogicalPlan {
           uniqueInputs.map {
             case r: ResolvedAttribute =>
               val dupAttrs = dups.collect { case x: ResolvedAttribute if x.name == r.name => x }
-              r.copy(sourceColumns = r.sourceColumns ++ dupAttrs.flatMap(_.sourceColumns))
+              if (dupAttrs.isEmpty) {
+                r
+              } else {
+                SingleColumn(MultiColumn(Seq(r) ++ dupAttrs, None, None), None, None, None)
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala:715]: No need to wrap by `SingleColumn` if there are no duplicated keys.

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala [hunk 7]
```diff
@@ -687,8 +691,15 @@ object LogicalPlan {
...
-    override def inputAttributes: Seq[Attribute] =
-      left.outputAttributes ++ right.outputAttributes
+    override def inputAttributes: Seq[Attribute] = {
+      (getRelationName(left) match {
+        case Some(q: String) => left.outputAttributes.map(_.withQualifier(q))
+        case None            => left.outputAttributes
+      }) ++ (getRelationName(right) match {
+        case Some(q: String) => right.outputAttributes.map(_.withQualifier(q))
+        case None            => right.outputAttributes
+      })
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala:701]: Propagate relation names for resolving columns that have a table name as prefix.

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/Expression.scala [hunk 8]
```diff
@@ -313,23 +313,40 @@ object Expression {
...
     private def matchesWithMultiColumn(columnName: String): Boolean = {
       expr match {
-        case MultiColumn(inputs, _) =>
-          inputs.collectFirst { case r: ResolvedAttribute => r }.exists(_.name == columnName)
+        case MultiColumn(inputs, name, _) =>
+          name.contains(columnName) || inputs.exists {
+            case r: ResolvedAttribute => r.name == columnName
+            case s: SingleColumn      => s.matchesWith(columnName)
+            case _                    => false
+          }
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/Expression.scala:342]: Fixed matching logic for `MultiColumn`

File: airframe-sql/src/main/scala/wvlet/airframe/sql/analyzer/TypeResolver.scala [hunk 9]
```diff
@@ -336,6 +335,8 @@ object TypeResolver extends LogSupport {
       case _ =>
         List(expr)
     }
+    trace(s"findMatchInInputAttributes: ${expr}, inputAttributes: ${inputAttributes} -> ${results}")
```
Comment (takezoe) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/analyzer/TypeResolver.scala:338]: Moved from the head of the method to here in order to include lookup results in the log message.

File: airframe-sql/src/test/scala/wvlet/airframe/sql/analyzer/TypeResolverTest.scala [hunk 10]
```diff
@@ -155,6 +178,14 @@ class TypeResolverTest extends AirSpec {
...
   test("resolve select from values") {
+    /**
+     * TODO This pattern cannot be round-tripped yet due to insufficient AllColumns handling.
+     * [generated sql from resolved plan]
+     * SELECT
+     *   MultiColumn(List(Literal(1), Literal(2), Literal(3)),None,None) AS id,
+     *   MultiColumn(List(Literal('one'), Literal('two'), Literal('three')),None,None) AS name
+     * FROM ((VALUES (1, 'one'), (2, 'two'), (3, 'three'))) AS t(id, name)
+     */
     val p = analyze("SELECT * FROM (VALUES (1, 'one'), (2, 'two'), (3, 'three')) AS t (id, name)")
```
Comment (takezoe) [inline airframe-sql/src/test/scala/wvlet/airframe/sql/analyzer/TypeResolverTest.scala:189]: This pattern will be covered in another PR as a part of `AllColumns` handling improvement.

**Review by xerial — APPROVED** [review]
ok. LGTM

File: airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala (hunk 3 above)
Comment (xerial) [inline airframe-sql/src/main/scala/wvlet/airframe/sql/model/LogicalPlan.scala:829]: Let's add a unique column name generator later.
